select union subquery NullPointerException

### Which version of ShardingSphere did you use?
<dependency>
            <groupId>org.apache.shardingsphere</groupId>
            <artifactId>shardingsphere-jdbc-core-spring-boot-starter</artifactId>
            <version>5.1.1</version>
</dependency>

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-JDBC

### Actual behavior
`Caused by: java.lang.NullPointerException: null
	at org.apache.shardingsphere.infra.binder.segment.select.subquery.engine.SubqueryTableContextEngine.createSubqueryTableContexts(SubqueryTableContextEngine.java:43)
	at org.apache.shardingsphere.infra.binder.segment.table.TablesContext.createSubqueryTables(TablesContext.java:99)
	at org.apache.shardingsphere.infra.binder.segment.table.TablesContext.<init>(TablesContext.java:86)
	at org.apache.shardingsphere.infra.binder.statement.dml.SelectStatementContext.<init>(SelectStatementContext.java:110)
	at org.apache.shardingsphere.infra.binder.SQLStatementContextFactory.getDMLStatementContext(SQLStatementContextFactory.java:146)
	at org.apache.shardingsphere.infra.binder.SQLStatementContextFactory.newInstance(SQLStatementContextFactory.java:129)
	at org.apache.shardingsphere.infra.binder.SQLStatementContextFactory.newInstance(SQLStatementContextFactory.java:114)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.<init>(ShardingSpherePreparedStatement.java:181)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.<init>(ShardingSpherePreparedStatement.java:149)
	at org.apache.shardingsphere.driver.jdbc.core.connection.ShardingSphereConnection.prepareStatement(ShardingSphereConnection.java:80)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.instantiateStatement(PreparedStatementHandler.java:86)
	at org.apache.ibatis.executor.statement.BaseStatementHandler.prepare(BaseStatementHandler.java:88)`

### Reason analyze (If you can)
Suspected to have something to do with the 5.1.0 update _Support select union/union all ... statements by federation engine_
5.0.0 no problem

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
`SELECT
              af.id,
              af.appid,
              af.uuid,
              af.wallet_id,
              af.currency,
              af.existing_balance,
              af.original_balance,
              af.balance_change,
              af.create_time,
              af.micro_time,
              af.operation_type,
              af.operation_mode
          FROM
              api_flow af
          WHERE
              af.create_time &gt;= #{startTime}
              AND af.create_time &lt; #{endTime}
          UNION
          SELECT
              af.id,
              af.appid,
              af.uuid,
              af.wallet_id,
              af.currency,
              af.existing_balance,
              af.original_balance,
              af.balance_change,
              af.create_time,
              af.micro_time,
              af.operation_type,
              af.operation_mode
          FROM
              ( SELECT DISTINCT wallet_id FROM api_flow WHERE create_time >= #{startTime} AND create_time &lt; #{endTime} ) t
              INNER JOIN api_fund_check_end cf ON t.wallet_id = cf.wallet_id
              INNER JOIN api_flow af ON ( cf.wallet_id = af.wallet_id AND cf.micro_time = af.micro_time)`

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SubqueryExtractUtil.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/SubqueryExtractUtil.java
@@ -20,6 +20,7 @@ package org.apache.shardingsphere.sql.parser.sql.common.util;
 import lombok.AccessLevel;
 import lombok.NoArgsConstructor;
 import org.apache.shardingsphere.sql.parser.sql.common.constant.SubqueryType;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.combine.CombineSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BetweenExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BinaryOperationExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.ExistsSubqueryExpression;
@@ -63,6 +64,13 @@ public final class SubqueryExtractUtil {
         extractSubquerySegmentsFromTableSegment(result, selectStatement.getFrom());
         if (selectStatement.getWhere().isPresent()) {
             extractSubquerySegmentsFromExpression(result, selectStatement.getWhere().get().getExpr());
+        }
+        extractSubquerySegmentsFromCombines(result, selectStatement.getCombines());
+    }
+    
+    private static void extractSubquerySegmentsFromCombines(final List<SubquerySegment> result, final Collection<CombineSegment> combineSegments) {
+        for (CombineSegment each : combineSegments) {
+            extractSubquerySegments(result, each.getSelectStatement());
         }
     }
     

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/util/SubqueryExtractUtilTest.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/util/SubqueryExtractUtilTest.java
@@ -18,7 +18,9 @@
 package org.apache.shardingsphere.sql.parser.sql.common.util;
 
 import org.apache.shardingsphere.sql.parser.sql.common.constant.AggregationType;
+import org.apache.shardingsphere.sql.parser.sql.common.constant.CombineType;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.column.ColumnSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.combine.CombineSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BinaryOperationExpression;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.ExpressionSegment;
 import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.InExpression;
@@ -163,5 +165,21 @@ public final class SubqueryExtractUtilTest {
         selectStatement.setWhere(new WhereSegment(0, 0, new InExpression(0, 0,
                 left, new SubqueryExpressionSegment(new SubquerySegment(0, 0, new MySQLSelectStatement())), false)));
         return new SubquerySegment(0, 0, selectStatement);
+    }
+    
+    @Test
+    public void assertGetSubquerySegmentsWithCombineSegment() {
+        SelectStatement selectStatement = new MySQLSelectStatement();
+        selectStatement.getCombines().add(new CombineSegment(0, 0, CombineType.UNION, createSelectStatementForCombineSegment()));
+        Collection<SubquerySegment> actual = SubqueryExtractUtil.getSubquerySegments(selectStatement);
+        assertThat(actual.size(), is(1));
+    }
+    
+    private SelectStatement createSelectStatementForCombineSegment() {
+        SelectStatement result = new MySQLSelectStatement();
+        ExpressionSegment left = new ColumnSegment(0, 0, new IdentifierValue("order_id"));
+        result.setWhere(new WhereSegment(0, 0, new InExpression(0, 0,
+                left, new SubqueryExpressionSegment(new SubquerySegment(0, 0, new MySQLSelectStatement())), false)));
+        return result;
     }
 }

